designs/aws/aws-launch-templates-v2.md

```diff
@@ -0,0 +1,146 @@
+```
+kind: Provisioner
+spec:
+  provider:
+    launchTemplate: CustomKarpenterLaunchTemplateDemo
+     autofill: true
+```
+
+*Pros*
+* A smaller launch template is much easier to maintain for the user.
+* Our API shape remains clean and easier to use. This paradigm is well known across other AWS capacity providers like MNG, Batch etc.
+* If someone wants to customize the entire LaunchTemplate, they can continue to do so and specify `autofill: false`
+
+*Cons*
+* Our validation logic will start to look very messy due to odd scenarios.
+Example - if someone gives us a partial launch template with a custom AMI, the UserData cannot be autofilled by Karpenter.
+Our API shape remains clean and easier to use. This paradigm is well known across other AWS capacity providers like MNG, Batch etc.
+* The user still needs to maintain a launch template. For some, the pain of maintaining a launch template is invariant to how many fields they have to fill out.
+
+
+
+**Approach 3 - Provide a more generic way to specify launch template contents in the Provider**
+
+The Provider today is strongly typed and we only support a few fields. We can introduce a new field like launchTemplateOverride where we accept bytes and attempt to do deserialize it into a launch template field at runtime.
+
+Effectively, a customer writes the provider as such -
+
+```
+kind: Provisioner
+spec:
+  provider:
+    launchTemplateOverride:
+        BlockDeviceMappings:
+        - DeviceName: "/dev/xvdb"
+        Ebs:
+            Iops: 3000
+            VolumeSize: 30
+            VolumeType: gp3
+```
+
+and everything under launchTemplateOverride we try and deserialize [to the LaunchTemplateData](https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RequestLaunchTemplateData.html) object using some form of reflection.
+
+
+*Pros*
+* You get the full ability of EC2 launch template right within your Provider
+* You don't need to maintain a launch template at all. Any new EC2 feature will work out of the box as long as the deserialization logic uses the latest AWS SDK i.e keep the controller up to date.
+* It most directly addresses what every customer is asking - let them configure EC2 properties from within the provisioner.
+
+*Cons*
+* APIs should really never accept weakly typed input. It's going to be very easy for a customer to get something wrong what went wrong break the deserialization. When that happens, it might be equally difficult to debug YAML.
+* If customers have to fill out the whole launch templates, it's going to make the provisionerSpec verbose. If they only fill it out partially, then you have the same cons as approach 2.
+
+
+
```

Comment:
The 4th approach we talked about at the community meeting is specifying more parameters in the podspec as `karpenter.sh` namespaced labels. Can we explore this more?

Reply:
I'll add that in. @ellistarn that's pretty much what we discussed at the last working group. It's nice to have a separation of responsibilities and have the application developer only define what their application requires and not have entire control over the actual infrastructure. It'd make it much harder to enforce things like "all our nodes should be using AMIs XX with Security Tooling YY" if every app developer needs to do per podSpec.